streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java

```diff
@@ -175,6 +175,12 @@ private void setState(final State newState) {
             }
 
             state = newState;
+
+            // Question: Is this a good idea? Or should we spread the latch countdown to finally blocks
+            // Count down the latch if transitioning from CREATED to any other state
+            if (oldState == State.CREATED && newState != State.CREATED) {
+                initializationLatch.countDown();
+            }
```

Comment:
@gharris1727 This works (all the tests pass), but I want to double-check it with you. The other alternative would be to count down in `finally` blocks across the class.

Reply:
> The other alternative would be to count down in finally blocks across the class.

It could make sense to have a finally block in the `initialize` method and call `countDown` in `shutdown`, but the current implementation appears to have reasonable semantics.

I have a weak feeling that counting down this latch is too much for the setState method to be doing, but at the same time pushing the latch management out into the calling code risks missing a (possibly future) call-site which transitions away from CREATED.

I think i would leave this to see what the Streams folks think about it.

Reply:
I agree with @gharris1727 -- this is not the right place for this code. I am not 100% sure right now what the best place would be, but both `run()` or `initialize()` look like good candidates (maybe `initialize()` is better).

Reply:
I moved the `countDown` logic to the `finally` block of `initialize`. It sounds reasonable to me.